### ISSUE TITLE
Jobs that don't finish within the timeout period might send duplicate notifications

### ISSUE THREAD

**Issue by jimleroyer** [cds-snc/notification-planning#79]
### Task Description

Every minute, [a celery task wakes up](https://github.com/cds-snc/notification-api/blob/e1eaf3511b42dee051eac3c0b097836a6f63375e/app/celery/tasks.py#L595) to check the process of unfinished job. [If the job took more than one hour](https://github.com/cds-snc/notification-api/pull/1375/files), then it restart the job and all remaining notifications to be sent.

The issue arises when the job has not finished because concurrent sending of notifications slowed it down. The job still runs but it is marked as incomplete and hence, restart the same job that will send notifications. We can detect these by comparing the `started_at` and `processing_started` fields: the latter usually starts at the same time as the former, but in the case of a job reset, it will be updated one hour late at the exact `00` second mark as it was started by an automated process.

To fix this, one avenue is to not have a timeout duration for jobs on the starting time, but rather on the latest notification sent. The following changes might be necessary:

1. Change the celery check-job-status job to rely on the `updated_at` field instead of `started_at` to detect if the job hangs.
2. Update the `updated_at` field through a new celery job that would trigger every minute via the run-scheduled-jobs job. The latter should kick off the new job and not have the job ran inside, i.e. the same as `process_job` but for updating job instead (e.g. `update_job`). 
3. Within the new `update_job` celery job, update the `updated_at` field by looking at the latest updated notifications associated with the given job and copy the `updated_at` timestamp of the most recent notifications to the job column.

### Task Outcome

No more duplicates of emails after the job timeout period has passed.
No more negative numbers in the sending dashboards

### Acceptance Criteria

- [ ] Jobs that take longer than 1 minute to fully process have their `updated_at` field updated mid processing
- [ ] When Notify is under load, partially processed jobs that exceed the timeout period of 1 hour are not fully re-processed but instead begin processing only the notifications that have not already been sent.
- [ ] Partially processed jobs do not result in duplicate notification sends when being re-processed by another Celery worker

### QA steps
1. Open Blazer and query the jobs for the service you will perform testing on
2. Do a large bulk send either via the API or the UI, using a template with a large amount of content
3. Note that the job's `updated_at` column correlates with the `updated_at` column of the most recently added notification in the DB for that job
    - This may vary slightly, it may not always be the latest added notification as the task that updates the field in the job is periodic, so processing may complete before the job status is checked and the field updated. It should however always match up with the `updated_at` of one of the more recent notifications for that job.
4. Once the job has finished processing, check that the number of notifications inserted into the DB matches the `notification_count` for that job.
5. Check in admin that the sent notifications count matches the `notification_count` for that job.

As this work changes the behaviour of our job retry mechanisms, ideally we want to perform a number of sends in parallel that introduces enough of a slow down, such that a job does not complete before timing out (30 minutes) and needs to be retried. This may not be feasible at the QA level, but if it is, it should be attempted.

**Comment by yaelberger-commits** [comment]
Steve suggests bumping time up to 2 hours from current 1 hour as a quick fix for now

**Comment by yaelberger-commits** [comment]
Ideally bump now back to 1 hour and implement the fix above

**Comment by jimleroyer** [comment]
@yaelberger-commits We should probably put this one back in the backlog, I believe this is still current. We can talk about it if you need more details.

**Comment by jzbahrai** [comment]
two notification IDs to investigate: 
8b49f1d3-09b8-48b9-813f-90d623e6815d
5fd18da1-82b1-4084-8224-1c804089dca1

it seems we are calling "persist_notification"
 on failure? 
![Screenshot 2023-06-01 at 1.36.22 PM.png](https://images.zenhubusercontent.com/6149fdc10c14631b6b9c16dc/b69e1313-b363-4b46-936b-6ea84dd1fdab)![Screenshot 2023-06-01 at 1.36.38 PM.png](https://images.zenhubusercontent.com/6149fdc10c14631b6b9c16dc/41d3ad71-36cb-4536-89f6-0219b72f0b78)![Screenshot 2023-06-01 at 12.06.55 PM.png](https://images.zenhubusercontent.com/6149fdc10c14631b6b9c16dc/80f7007b-d1b5-4c00-9553-ea1b2f626613)

**Comment by jzbahrai** [comment]
for the same job_id and information "a line in the csv" - we are creating the same notification twice.

**Comment by whabanks** [comment]
Implemented the suggested solution of basing the job timeout on the time of the latest sent notification instead of the start time of the job processing. Looking into writing test cases that introduce delays when processing a job to test this pre-staging.

**Comment by smcmurtry** [comment]
After a fair bit of difficulty getting things set up right, I was able to test this out. I can't reproduce the original bug where we get duplicate notifications. And I found a new bug where jobs are being marked as complete before any notifications are actually sent (see here for a description: https://github.com/cds-snc/notification-api/pull/1912#issuecomment-1614678906).

Since this PR modifies a task that checks for incomplete jobs, it won't do anything to fix the duplicate notification issue.

**Comment by smcmurtry** [comment]
We should have a discussion about whether to keep working on this or not.

**Comment by whabanks** [comment]
[Initial solution was reverted](https://github.com/cds-snc/notification-api/pull/1929) as an [issue was identified](https://gcdigital.slack.com/archives/CV38DBNVA/p1689248476178229) with it.

**Comment by whabanks** [comment]
Started work addressing the issue with the initial solution briefly on Thursday. Switched to working on [#1275](https://app.zenhub.com/workspaces/notify-planning-614b3ad91bc2030015ed22f5/issues/gh/cds-snc/notification-planning/1275) with Steve M. Will continue work on this today.

**Comment by smcmurtry** [comment]
Will's PR merged this morning. We ended up uncovering and fixing some related bugs.

**Comment by jzbahrai** [comment]
@jzbahrai will look at QA this
